lib/models/instrumentation.js

```diff
@@ -0,0 +1,277 @@
+'use strict';
+
+var fs = require('fs-extra');
+var chalk = require('chalk');
+var experiments = require('../experiments/');
+var tree = require('heimdalljs-tree');
+var utilsInstrumentation = require('../utilities/instrumentation');
+
+var vizEnabled = utilsInstrumentation.vizEnabled;
+var instrumentationEnabled = utilsInstrumentation.instrumentationEnabled;
+
+function _enableFSMonitorIfInstrumentationEnabled() {
+  var monitor;
+  if (instrumentationEnabled()) {
+    var FSMonitor = require('heimdalljs-fs-monitor');
+    monitor = new FSMonitor();
+    monitor.start();
+  }
+  return monitor;
+}
+
+_enableFSMonitorIfInstrumentationEnabled();
+
+
+function Instrumentation (options) {
+  this.ui = options.ui;
+
+  // project constructor will set up bidirectional link
+  this.project = null;
+
+  this.instrumentations = {
+    init: options.initInstrumentation,
+    build: {
+      cookie: null,
+      node: null,
+      count: 0,
+    },
+    command: {
+      cookie: null,
+      node: null,
+    },
+    shutdown: {
+      cookie: null,
+      node: null,
+    },
+  };
+
+  this._heimdall = null;
+
+  if (!options.initInstrumentation && this.isEnabled()) {
+    this.instrumentations.init = {
+      cookie: null,
+      node: null,
+    };
+    this.start('init');
+
+    this.ui.writeLine(chalk.yellow(
+      'No init instrumentation passed to CLI.  Please update your global ember or ' +
+      'invoke ember via the local executable within node_modules.  Init ' +
+      'instrumentation will still be recorded, but some bootstraping will be ' +
+      'omitted.'
+    ));
+  }
+}
+
+module.exports = Instrumentation;
+
+Instrumentation.prototype._buildSummary = function (tree, result, resultAnnotation) {
+  var buildSteps = 0;
+  var totalTime = 0;
+
+  var node;
+  var statName;
+  var statValue;
+  var nodeItr;
+  var statsItr;
+  var nextNode;
+  var nextStat;
+
+  for (nodeItr = tree.preOrderIterator();;) {
```

Comment:
I'd like to see this split out, it is a tad bit annoying to have to include the iterator twice, seems good for a future refactor (or just wait until we can use ES6?)...